File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        n = len(preorder)
        inDict = {j:i for i, j in enumerate(inorder)}
        self.pos = 0
        def dfs(left, right):
            if left > right:
                return None
            val = preorder[self.pos]
            self.pos += 1
            node = TreeNode(val)
            idx = inDict[val]
            node.left = dfs(left, idx - 1)
            node.right = dfs(idx + 1, right)

            return node
        
        return dfs(0, n - 1)


        

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        // there is no native queue in javascript so a pointer 
        // need to be maintained to track which element to remove.
        const queue = [[root, 0]]; // store the depth and the root. 
        let i = 0;
        const ans = [];
        while (i < queue.length) {
            const [node, depth] = queue[i++]; // post fix operator.
            if (!node) continue // to catch if the root node itself if null;
            if (ans.length > depth) {
                ans[depth].push(node.val);
            } else {
                ans.push([node.val]);
            }
            if (node.left) queue.push([node.left, depth + 1]);
            if (node.right) queue.push([node.right, depth + 1]);
        }
        return ans;
    }
}
